paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkTableSink.java

```diff
@@ -18,168 +18,110 @@
+        // Since only UPDATE_AFTER type messages can be received at present,
+        // AppendOnlyFileStoreTable and ChangelogValueCountFileStoreTable without primary keys
+        // cannot correctly handle old data, so they are marked as unsupported. Similarly, it is not
+        // allowed to update the primary key column when updating the column of
+        // ChangelogWithKeyFileStoreTable, because the old data cannot be handled correctly.
+        if (table instanceof ChangelogWithKeyFileStoreTable) {
             Options options = Options.fromMap(table.options());
-            if (options.get(CHANGELOG_PRODUCER) == ChangelogProducer.INPUT) {
-                return requestedMode;
-            }
-
-            if (options.get(MERGE_ENGINE) == MergeEngine.AGGREGATE) {
-                return requestedMode;
-            }
-
-            if (options.get(LOG_CHANGELOG_MODE) == LogChangelogMode.ALL) {
-                return requestedMode;
-            }
-
-            // with primary key, default sink upsert
-            ChangelogMode.Builder builder = ChangelogMode.newBuilder();
-            for (RowKind kind : requestedMode.getContainedKinds()) {
-                if (kind != RowKind.UPDATE_BEFORE) {
-                    builder.addContainedKind(kind);
+            Set<String> primaryKeys = new HashSet<>(table.primaryKeys());
+            updatedColumns.forEach(
+                    column -> {
+                        if (primaryKeys.contains(column.getName())) {
+                            throw new UnsupportedOperationException(
+                                    "Updates to primary keys are not supported.");
+                        }
+                    });
+
+            if (options.get(MERGE_ENGINE) == MergeEngine.DEDUPLICATE) {
+                return new RowLevelUpdateInfo() {};
+            } else if (options.get(MERGE_ENGINE) == MergeEngine.PARTIAL_UPDATE) {
+                this.updatedColumns = new LinkedHashMap<>();
+                for (int i = 0; i < columns.size(); i++) {
+                    Column column = columns.get(i);
+                    if (primaryKeys.contains(column.getName()) || updatedColumns.contains(column)) {
+                        this.updatedColumns.put(
+                                i,
+                                RowData.createFieldGetter(
+                                        column.getDataType().getLogicalType(), i));
+                    }
                 }
+                // Even with partial-update we still need all columns. Because the topology
+                // structure is source -> cal -> ConstraintEnforcer -> sink, in the
+                // constraintEnforcer operator, the constraint check will be performed according to
+                // the index, not according to the column. So we can't return only some columns,
+                // which will cause problems like ArrayIndexOutOfBoundsException.
+                return new RowLevelUpdateInfo() {};
```

Comment:
So if here we just require all columns, why we need `updatedColumns`?

Reply:
In the case of `partial-update`, we should only update the `primaryKeys ` and `updatedColumns`, and set the other columns to null.

Reply:
I just removal it, and I found that tests are also passed.
It is useless now, because here you just return all columns.

Reply:
I support remove `updatedColumns` too. In partial-update, replacing a row field with its old field value won't cause error. So I think we don't need to handle it in `RowDataStoreWriteOperator`.

Reply:
@JingsongLi @yuzelin Do we need to consider the case of concurrent writing? For example, the table has 5 columns <a, b, c, d, e>, one job-a is updating column a, and another job-b is updating b. If both jobs write all the columns, will there be an error when job-c reads all the columns?

Reply:
No, concurrent writing is a big thing, it is not only columns, but also rows.

Reply:
Okay, I will remove this logic.